smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/EventStreamGenerator.java

```diff
@@ -0,0 +1,518 @@
+        } else {
+            writer.write("return $L(output, context);", errorDeserMethodName);
+        }
+    }
+
+    // Parse members from event headers.
+    private void readEventHeaders(
+        GenerationContext context,
+        StructureShape event,
+        BiFunction<String, MemberShape, String> getEventHeaderOutputValue
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        List<MemberShape> headerMembers = event.getAllMembers().values().stream()
+                .filter(member -> member.hasTrait(EventHeaderTrait.class)).collect(Collectors.toList());
+        for (MemberShape headerMember : headerMembers) {
+            String memberName = headerMember.getMemberName();
+            writer.openBlock("if (output.headers[$S] !== undefined) {", "}", memberName, () -> {
+                String headerValue = getEventHeaderOutputValue.apply(
+                        "output.headers['" + memberName + "']", headerMember);
+                writer.write("contents.$L = $L;", memberName, headerValue);
+            });
+        }
+    }
+
+    private void readEventBody(
+        GenerationContext context,
+        StructureShape event,
+        Set<Shape> eventShapesToDeserialize
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        Shape payloadShape = getEventPayloadShape(context, event);
+        if (payloadShape instanceof BlobShape) {
+            // Since event itself must be a structure shape, so blob payload member must has eventPayload
+            // trait explicitly.
+            MemberShape payloadMember = event.getAllMembers().values().stream()
+                    .filter(member -> member.hasTrait(EventPayloadTrait.class))
+                    .collect(Collectors.toList()).get(0);
+            writer.write("contents.$L = output.body;", payloadMember.getMemberName());
+        } else if (payloadShape instanceof StringShape) {
+            // Since event itself must be a structure shape, so string payload member must has eventPayload
+            // trait explicitly.
+            MemberShape payloadMember = event.getAllMembers().values().stream()
+                    .filter(member -> member.hasTrait(EventPayloadTrait.class))
+                    .collect(Collectors.toList()).get(0);
+            writer.write("contents.$L = await collectBodyString(output.body, context);", payloadMember.getMemberName());
+        } else if (payloadShape instanceof StructureShape || payloadShape instanceof UnionShape) {
+            // If body is Structure or Union, they we need to parse the string into JavaScript object.
+            writer.write("const data: any = await parseBody(output.body, context);");
+            Symbol symbol = getSymbol(context, payloadShape);
+            String deserFunctionName = ProtocolGenerator.getDeserFunctionName(symbol, context.getProtocolName());
+            writer.write("contents = {...contents, ...$L(data, context)};", deserFunctionName);
```

Comment:
completely optional: if using `const contents`, this could be `Object.assign(contents, ...)`